src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -6373,6 +6405,354 @@ static void Main()
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y4").WithLocation(25, 9));
         }
 
+        [WorkItem(57517, "https://github.com/dotnet/roslyn/issues/57517")]
+        [Fact]
+        public void TypeInference_03()
+        {
+            var source =
+@"using System;
+using System.Linq.Expressions;
+
+void Test1<T>(Func<T> exp) {}
+void Test2<T>(Expression<Func<T>> exp) {}
+void Test3<T>(Func<Func<T>> exp) {}
+void Test4<T>(Func<Expression<Func<T>>> exp) {}
+
+Test1(() => 1);
+Test2(() => 2);
+Test3(() => () => 3);
+Test4(() => () => 4);
+";
+
+            var expectedDiagnostics = new[]
+            {
+                // (6,6): warning CS8321: The local function 'Test3' is declared but never used
+                // void Test3<T>(Func<Func<T>> exp) {}
+                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "Test3").WithArguments("Test3").WithLocation(6, 6),
+                // (7,6): warning CS8321: The local function 'Test4' is declared but never used
+                // void Test4<T>(Func<Expression<Func<T>>> exp) {}
+                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "Test4").WithArguments("Test4").WithLocation(7, 6),
+                // (11,1): error CS0411: The type arguments for method 'Test3<T>(Func<Func<T>>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                // Test3(() => () => 3);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "Test3").WithArguments("Test3<T>(System.Func<System.Func<T>>)").WithLocation(11, 1),
```

Comment:
I'm not familiar at all with the rules/specifications around this so feel free to ignore, but is it expected that this case shouldn't support inference? In my bug report (57517) I mentioned that `Func<Func<T>>` supported inference of `T` and it seems this change is going in the opposite direction and undoing something that was working? When I made the bug report I was thinking that the outcome would be that inference of `T` on `Func<Expression<Func<T>>>`  would start working.

Reply:
Thanks for catching this and for logging [https://github.com/dotnet/roslyn/issues/57517](https://github.com/dotnet/roslyn/issues/57517).

With C#10 currently, there is a potential breaking change from C#9 in type inference when inferring from the return type of a lambda expression _when the return value is a nested lambda expression or a method group_
(see examples in [https://github.com/dotnet/roslyn/issues/57517#issuecomment-964289629](https://github.com/dotnet/roslyn/issues/57517#issuecomment-964289629) and [https://github.com/dotnet/roslyn/issues/57630](https://github.com/dotnet/roslyn/issues/57630)).

We've decided to fix the breaking change for now by _not inferring from the return type of a lambda if the return value is another lambda or method group_. This effectively reverts the type inference behavior _for those cases_ to the C#9 behavior.

In your examples above, as you noted, this means that inference fails for the calls to `Test3()` and `Test4()`, as in C#9.

Separately, we will investigate whether we can improve type inference (in a subsequent release) for these cases by inferring from the _function type_ of the returned lambda or method group. If so, that should allow inference for `Test3()` and `Test4()`.

I will log a separate issue for that investigation.